File: api/src/main/java/com/timgroup/eventstore/api/legacy/ReadableLegacyStore.java
package com.timgroup.eventstore.api.legacy;

import java.util.function.LongFunction;
import java.util.function.ToLongFunction;
import java.util.stream.Stream;

import com.timgroup.eventstore.api.EventData;
import com.timgroup.eventstore.api.EventInStream;
import com.timgroup.eventstore.api.EventReader;
import com.timgroup.eventstore.api.EventStore;
import com.timgroup.eventstore.api.EventStream;
import com.timgroup.eventstore.api.EventStreamJavaAdapter;
import com.timgroup.eventstore.api.Position;
import com.timgroup.eventstore.api.ResolvedEvent;
import org.joda.time.DateTimeZone;
import scala.Function1;
import scala.Option;
import scala.collection.Seq;
import scala.runtime.BoxedUnit;

public final class ReadableLegacyStore implements EventStore {
    private final EventReader eventReader;
    private final LongFunction<Position> toPosition;
    private final ToLongFunction<Position> toVersion;

    public ReadableLegacyStore(EventReader eventReader, LongFunction<Position> toPosition, ToLongFunction<Position> toVersion) {
        this.eventReader = eventReader;
        this.toPosition = toPosition;
        this.toVersion = toVersion;
    }

    @Override
    public EventStream fromAll(long version) {
        return new EventStreamJavaAdapter(eventReader.readAllForwards(toPosition.apply(version)).map(this::toEventInStream));
    }

    @Override
    public long fromAll$default$1() {
        return 0;
    }

    @Override
    public void fromAll(long version, Function1<EventInStream, BoxedUnit> eventHandler) {
        try (Stream<ResolvedEvent> stream = eventReader.readAllForwards(toPosition.apply(version))) {
            stream.map(this::toEventInStream).forEachOrdered(eventHandler::apply);
        }
    }

    @Override
    public void save(Seq<EventData> newEvents, Option<Object> expectedVersion) {
        throw new UnsupportedOperationException();
    }

    @Override
    public Option<Object> save$default$2() {
        throw new UnsupportedOperationException();
    }

    private EventInStream toEventInStream(ResolvedEvent in) {
        return new EventInStream(
                new org.joda.time.DateTime(in.eventRecord().timestamp().toEpochMilli(), DateTimeZone.UTC),
                EventData.apply(in.eventRecord().eventType(), in.eventRecord().data()),
                toVersion.applyAsLong(in.position()));
    }

    @Override
    public String toString() {
        return eventReader.toString();
    }
}

File: api-legacy/src/main/scala/com/timgroup/eventstore/api/legacy/ReadableLegacyStore.scala
package com.timgroup.eventstore.api.legacy

import java.util.function.{LongFunction, ToLongFunction}
import java.util.stream

import com.timgroup.eventstore.api.{EventData, EventInStream, EventReader, EventStore, EventStream, Position, ResolvedEvent}

class ReadableLegacyStore(private val eventReader: EventReader, private val toPosition: LongFunction[_ <: Position], private val toVersion: ToLongFunction[_ >: Position]) extends EventStore {
  override def fromAll(version: Long): EventStream = new EventStream {
    private val underlying = eventReader.readAllForwards(toPosition.apply(version)).iterator()

    override def hasNext: Boolean = underlying.hasNext

    override def next(): EventInStream = toEventInStream(underlying.next())
  }

  override def streamingFromAll(version: Long): stream.Stream[EventInStream] = eventReader.readAllForwards(toPosition.apply(version)).map(new java.util.function.Function[ResolvedEvent, EventInStream] {
    override def apply(t: ResolvedEvent): EventInStream = toEventInStream(t)
  })

  override def save(newEvents: Seq[EventData], expectedVersion: Option[Long]): Unit = throw new UnsupportedOperationException()

  override def toString: String = eventReader.toString

  private def toEventInStream(in: ResolvedEvent) = EventInStream(
    new org.joda.time.DateTime(in.eventRecord().timestamp().toEpochMilli, org.joda.time.DateTimeZone.UTC),
    EventData(in.eventRecord().eventType(), in.eventRecord().data()),
    toVersion.applyAsLong(in.position())
  )
}
